### ISSUE TITLE
Python default arguments

### ISSUE THREAD

**Issue by swails** [swig/swig#387]
This is very similar to #377.

In one of the projects I work on, SWIG is used to create Python bindings for a C++-based code.  There is some custom Python code added to each method that makes use of the assumption that the call signature looks like ``def func(*args):`` rather than ``def func(arg1, arg2, ...)``.  For example, using SWIG 2.0.5, I'll see SWIG-generated Python code like this:

```Python
    def addForce(self, *args):
        """ Some docstring """
        # This Python code is added via pythonprepend_all.i
        if not args[0].thisown:
           s = "the %s object does not own its corresponding object" \
              % args[0].__class__.__name__
           raise Exception(s)


        val = _openmm.System_addForce(self, *args)
        args[0].thisown=0


        return val
```

As of SWIG 3.0.5, all of these methods are replaced by the actual arguments that correspond to the prototype call signature:

```Python
    def addForce(self, force):
        """ Some docstring """
        # This Python code is added via pythonprepend_all.i
        if not args[0].thisown:
           s = "the %s object does not own its corresponding OpenMM object" \
              % args[0].__class__.__name__
           raise Exception(s)


        val = _openmm.System_addForce(self, force)
        args[0].thisown=0


        return val
```

Obviously this latter one introduces a `NameError` or `UnboundLocalError` as `args` is not defined.  This is an open issue in that project: pandegroup/OpenMM#879.

Is there any way to go back to the old behavior of having every wrapper accept `*args`?  Is this an option whose default option was changed from SWIG 2 to SWIG 3?

**Comment by swails** [comment]
`git-bisect` traced the regression to fdc6bbeda387fb49ab7dcdbb0f17e4645db3115c (and the log message certainly suggests this was an intended change).

The question of how to restore the original behavior still stands, though.

**Comment by vadz** [comment]
Yes, it was intentional, I'm pretty sure that the original behaviour was just a bug (and SWIG hadn't behaved like this before Python 3 changes) and it's definitely very bad from the point of view of actually using the generated Python API as there is no indication whatsoever about which arguments the functions take.

Unfortunately fixing this bug exposed quite a few other bugs in the nearby code. Some of them were already fixed, others, such as #377 and #327 still remain and I hope to finally have a look at them soon.

But in your case it's not possible to fix it in SWIG at all as you rely on having these `*args`, so the only thing I can suggest is to use `compactdefaultargs` feature globally. Normally this should re-enable the old behaviour. If this works for you, please close this issue as I don't see what else could be done here.

**Comment by swails** [comment]
Very simple!  Thanks!
